actionSHARK/utils.py

```diff
@@ -0,0 +1,148 @@
+from typing import Optional, Any, Tuple, List
+from mongoengine import ObjectIdField
+from pycoshark.mongomodels import Commit, JobStep, RunPullRequest
+import sys
+import datetime as dt
+
+
+def parse_date(value: Optional[str] = None, is_millisecond: bool = False
+) -> Optional[dt.datetime]:
+    """Convert Datetime string to actual Datetime object.
+
+    Args:
+        value (str): The datetime string to be converted. Defaults to None.
+        is_millisecond (bool, optional): If datetime has millisecond in the string. Defaults to False.
+
+    Returns:
+        datetime: datetime formated or None if no string passed.
+    """
+    if not value:
+        return None
+
+    date_fmt = "%Y-%m-%dT%H:%M:%S%z"
+    if is_millisecond:
+        date_fmt = "%Y-%m-%dT%H:%M:%S.%f%z"
+
+    if sys.version_info.minor < 8:
+        date_fmt = date_fmt.replace("%z", "Z")
+    return dt.datetime.strptime(value, date_fmt)
+
+
+def to_int(value: Optional[str] = None) -> Optional[int]:
+    """
+    Converting string to integer.
+
+    Args:
+        value (str, optional): The value to convert. Defaults to None.
+
+    Returns:
+        Optional[int]
+    """
+
+    if not value:
+        return None
+
+    return int(value)
+
+
+def format_repository_url(value: Optional[str] = None) -> Optional[str]:
+
+    if not value:
+        return None
+
+    value = value.replace("api.", "")
+    value = value.replace("repos/", "")
+    value = value.replace(".git", "")
+    value = value + ".git"
+
+    return value
+
+
+def run_head_repository_url(value: str) -> str:
+    """
+    Construct URL from repository full name "owner/repository"
+
+    Args:
+        value (str): Foramt "owner/repository"
+
+    Returns:
+        str: GitHub .git url
+    """
+    return "https://github.com/" + value + ".git"
+
+
+def commit_object_id( value: Optional[str] = None) -> Optional[ObjectIdField]:
+    """
+    Find Run object_id from Run collection.
+
+    Args:
+        value (int): Run id.
+
+    Returns:
+        Optional[ObjectIdField]
+    """
+
+    if not value:
+        return None
+
+    try:
+        r = Commit.objects.get(revision_hash=value).id
```

Comment:
Agreed on using project ID instead of VCS system ID. However, the commit module lacks a project ID field, making it challenging to retrieve information. Planning to revisit commit-related fields after completing vcsShark.

Reply:
you would first get the vcs_system via the project then use vcs_system_id

Reply:
But it in our new design we have multiple VcsSystems for each project. So it will return multiple vcsSystems.

Reply:
You would assume the latest vcs_system_id would be the correct one in that case so use that.